### PR TITLE
Custom functions guide improvements

### LINKED ISSUE #1123: Custom functions guide improvements
### Description

Elaborate more on:
- various argument types: https://github.com/handsontable/hyperformula/pull/1108#discussion_r1052724327
- `SimpleRangeValue` class: https://github.com/handsontable/hyperformula/pull/1108#discussion_r1052733577
- `runFunction` and automatic validation: https://github.com/handsontable/hyperformula/pull/1108#discussion_r1052736003

After completing this task, we will request the review of the custom functions guide from the external consultant

### Links

https://github.com/handsontable/hyperformula/issues/779
https://github.com/handsontable/hyperformula/pull/1108

### Type of issue

- [ ] Missing documentation
- [ ] Error in the documentation
- [ ] Request for a demo
- [ ] Request for a tutorial, guide etc.
- [x] Other issues, suggestions or ideas

### PR DESCRIPTION
**Pull request by sequba**
### Context

- [x] describe the possible function argument types
- [x] refactor the `runFunction` method completely
- [x] describe what the `runFunction` method does
- [x] describe `SimpleRangeValue` class in the API reference and link it to the custom-functions guide

### How did you test your changes?

- [x] unit tests passed
- [x] local docs build
- [x] Grammarly  spellcheck

### Types of changes

- [ ] Breaking change (a fix or a feature because of which an existing functionality doesn't work as expected anymore)
- [ ] New feature or improvement (a non-breaking change that adds functionality)
- [ ] Bug fix (a non-breaking change that fixes an issue)
- [ ] Additional language file, or a change to an existing language file (translations)
- [x] Change to the documentation

### Related issues:
Fixes #1123

### Checklist:

- [x] My change is compliant with the [OpenDocument](https://docs.oasis-open.org/office/OpenDocument/v1.3/os/part4-formula/OpenDocument-v1.3-os-part4-formula.html) standard.
- [x] My change is compatible with Microsoft Excel.
- [x] My change is compatible with Google Sheets.
- [x] My code follows the code style of this project.
- [ ] I described my changes in the [CHANGELOG.md](https://github.com/handsontable/hyperformula/blob/master/CHANGELOG.md) file.
- [ ] My changes require a documentation update.
- [ ] My changes require a migration guide.

### CODE REVIEW

File: docs/guide/built-in-functions.md [hunk 1]
```diff
@@ -226,7 +226,7 @@ Total number of functions: **{{ $page.functionsCount }}**
 | FORMULATEXT | Returns a formula in a given cell as a string.                                                                      | FORMULATEXT(Reference)                              |
 | HLOOKUP     | Searches horizontally with reference to adjacent cells to the bottom.                                               | HLOOKUP(Search_Criterion, Array, Index, Sort_Order) |
 | INDEX       | Returns the contents of a cell specified by row and column number. The column number is optional and defaults to 1. | INDEX(Range, Row [, Column])                        |
-| MATCH       | Returns the relative position of an item in an array that matches a specified value.                                | MATCH(Searchcriterion, Lookuparray, Type)           |
+| MATCH       | Returns the relative position of an item in an array that matches a specified value.                                | MATCH(Searchcriterion, Lookuparray, MatchType)      |
```
Comment (sequba) [inline docs/guide/built-in-functions.md:229]: This is unrelated to this task. See: https://github.com/handsontable/hyperformula/discussions/1148#discussioncomment-4759685

File: src/interpreter/plugin/BooleanPlugin.ts [hunk 2]
```diff
@@ -142,7 +142,7 @@ export class BooleanPlugin extends FunctionPlugin implements FunctionPluginTypec
    */
   public and(ast: ProcedureAst, state: InterpreterState): InterpreterValue {
     return this.runFunction(ast.args, state, this.metadata('AND'),
-      (...args) => !args.some((arg: boolean) => !arg)
+      (...args) => args.filter((arg: boolean | undefined) => arg !== undefined).every((arg: boolean) => !!arg)
```
Comment (sequba) [inline src/interpreter/plugin/BooleanPlugin.ts:145]: TODO: test it

File: src/interpreter/plugin/BooleanPlugin.ts [hunk 3]
```diff
@@ -156,7 +156,7 @@ export class BooleanPlugin extends FunctionPlugin implements FunctionPluginTypec
    */
   public or(ast: ProcedureAst, state: InterpreterState): InterpreterValue {
     return this.runFunction(ast.args, state, this.metadata('OR'),
-      (...args) => args.some((arg: boolean) => arg)
+      (...args) => args.filter((arg: boolean | undefined) => arg !== undefined).some((arg: boolean) => arg)
```
Comment (sequba) [inline src/interpreter/plugin/BooleanPlugin.ts:159]: TODO: test it

File: src/interpreter/plugin/FunctionPlugin.ts [hunk 4]
```diff
@@ -355,99 +355,171 @@ export abstract class FunctionPlugin implements FunctionPluginTypecheck<Function
...
-        if (val instanceof SimpleRangeValue && argumentDefinitions[i].argumentType !== FunctionArgumentType.RANGE && argumentDefinitions[i].argumentType !== FunctionArgumentType.ANY) {
-          maxHeight = Math.max(maxHeight, val.height())
-          maxWidth = Math.max(maxWidth, val.width())
+    const resultArray: InternalScalarValue[][] = [ ...Array(resultArrayHeight).keys() ].map(row =>
+      [ ...Array(resultArrayWidth).keys() ].map(col => {
+        const vectorizedArguments = this.vectorizeAndBroadcastArgumentsIfNecessary(isVectorizationOn, argumentValues, argumentMetadata, row, col)
+        const result = this.calculateSingleCellOfResultArray(state, vectorizedArguments, argumentMetadata, argumentIgnorableFlags, functionImplementation, metadata.returnNumberType)
+
+        if (result instanceof SimpleRangeValue) {
+          throw 'Function returning array cannot be vectorized.' // TODO: test it
```
Comment (sequba) [inline src/interpreter/plugin/FunctionPlugin.ts:399]: TODO: test it

File: src/interpreter/plugin/BooleanPlugin.ts (hunk 2 above)
Comment (sequba) [reply src/interpreter/plugin/BooleanPlugin.ts:145]: done

File: src/interpreter/plugin/BooleanPlugin.ts (hunk 3 above)
Comment (sequba) [reply src/interpreter/plugin/BooleanPlugin.ts:159]: done

File: src/interpreter/plugin/BooleanPlugin.ts [hunk 5]
```diff
@@ -167,7 +167,7 @@ export class BooleanPlugin extends FunctionPlugin implements FunctionPluginTypec
   public xor(ast: ProcedureAst, state: InterpreterState): InterpreterValue {
     return this.runFunction(ast.args, state, this.metadata('XOR'), (...args) => {
       let cnt = 0
-      args.forEach((arg: boolean) => {
+      args.filter((arg: boolean | undefined) => arg !== undefined).forEach((arg: boolean) => {
```
[the remark this answers is not in the document]
Comment (sequba) [reply src/interpreter/plugin/BooleanPlugin.ts:170]: done

File: src/interpreter/plugin/FunctionPlugin.ts (hunk 4 above)
Comment (sequba) [reply src/interpreter/plugin/FunctionPlugin.ts:399]: done

**Review by adamu-handsontable — COMMENTED** [review]
I've checked it and left feedback

File: docs/guide/built-in-functions.md (hunk 1 above)
Comment (sequba) [reply docs/guide/built-in-functions.md]: This description follows the general assumption that _the functions work the same as in Microsoft Excel_. I don't like this assumption, but currently, [we do not have space for a more thorough documentation of the built-in functions](https://github.com/handsontable/hyperformula/issues/936).

So:
- MatchType accepts the same values as in [the Microsoft Excel implementation of `MATCH`](https://support.microsoft.com/en-us/office/match-function-e8dffd45-c762-47d6-bf89-533f4a37673a)
- it is indeed optional and I marked it as such in the documentation

File: src/interpreter/plugin/BooleanPlugin.ts (hunk 2 above)
Comment (sequba) [reply src/interpreter/plugin/BooleanPlugin.ts:145]: It's a bit more complicated. My refactor of `FunctionPlugin` changed the way strings are coerced to boolean values, which implicitly changed the behavior of boolean functions (e.g. `AND`). For this reason, this tiny modification in `BooleanPlugin` was necessary to keep the current behavior.

**Review by adamu-handsontable — APPROVED** [review]
Everything looks good, thank you
